src/qp/operators/BlockNestedJoin.java

```diff
@@ -0,0 +1,229 @@
+
+    static int filenum = 0;         // To get unique filenum for this operation
+    int batchsize;                  // Number of tuples per out batch
+    int outerBlockSize;
+    ArrayList<Integer> leftindex;   // Indices of the join attributes in left table
+    ArrayList<Integer> rightindex;  // Indices of the join attributes in right table
+    String rfname;                  // The file name where the right table is materialized
+    Batch outbatch;                 // Buffer page for output
+    Batch leftbatch;                // Buffer page for left input stream
+    Batch rightbatch;               // Buffer page for right input stream
+    ArrayList<Batch> outerBlock;    // Outer block for tuples
+    ObjectInputStream in;           // File pointer to the right hand materialized file
+
+    int lcurs;                      // Cursor for left side buffer
+    int rcurs;                      // Cursor for right side buffer
+    int bcurs;                      // Cursor for block
+    boolean eosl;                   // Whether end of stream (left table) is reached
+    boolean eosr;                   // Whether end of stream (right table) is reached
+    boolean eobo;                   // Whether end of block object
+
+    public BlockNestedJoin(Join jn) {
+        super(jn.getLeft(), jn.getRight(), jn.getConditionList(), jn.getOpType());
+        schema = jn.getSchema();
+        jointype = jn.getJoinType();
+        numBuff = jn.getNumBuff();
+    }
+
+    /**
+     * During open finds the index of the join attributes
+     * * Materializes the right hand side into a file
+     * * Opens the connections
+     **/
+    public boolean open() {
+        /** select number of tuples per batch **/
+        int tuplesize = schema.getTupleSize();
+        batchsize = Batch.getPageSize() / tuplesize;
+
+        outerBlockSize = numBuff - 2;
+
+        outerBlock = new ArrayList<>(outerBlockSize);
+
+        /** find indices attributes of join conditions **/
+        leftindex = new ArrayList<>();
+        rightindex = new ArrayList<>();
+        for (Condition con : conditionList) {
+            Attribute leftattr = con.getLhs();
+            Attribute rightattr = (Attribute) con.getRhs();
+            leftindex.add(left.getSchema().indexOf(leftattr));
+            rightindex.add(right.getSchema().indexOf(rightattr));
+        }
+        Batch rightpage;
+
+        /** initialize the cursors of input buffers **/
+        lcurs = 0;
+        rcurs = 0;
+        bcurs = 0;
+        eobo = false;
+        eosl = false;
+        /** because right stream is to be repetitively scanned
+         ** if it reached end, we have to start new scan
+         **/
+        eosr = true;
+
+        /** Right hand side table is to be materialized
+         ** for the Nested join to perform
+         **/
+        if (!right.open()) {
+            return false;
+        } else {
+            /** If the right operator is not a base table then
```

Comment:
What does this comment mean? So if the right relation is a "base table", it wouldn't need to be written out as a file?